src/firebase/CRUDServices/imageService.js

```diff
@@ -0,0 +1,117 @@
+import firebase from '../firebase';
+const { PubSub } = require('@google-cloud/pubsub');
+
+export default class ImageService {
+  constructor() {
+    self.storageRef = firebase.storage().ref();
+    self.sieRawImagesBucketRef = storageRef.child('sie-raw-images');
+    self.sieProcessedImagesBucketRef = storageRef.child('sie-processed-images');
+    self.subscriptionName =
+  'projects/cs6510-spr2021/subscriptions/sie-image-processing-result-test-sub';
+  }
+
+    // singleton instance.
+    static sieInstance = null;
+
+    /**
+     * creating singleton instance.
+     * @return {ImageService} instance
+     */
+    static getInstance = () => {
+      if (self.sieInstance == null) {
+        self.sieInstance = new ImageService();
+      }
+      return self.sieInstance;
+    }
+
+    /**
+     * Put a self image into the cloud storage.
+     * @param {String} userName user name.
+     * @param {String} experimentId experiment id.
+     * @param {File} image self image file.
+     * @return {Boolean} succeed or not.
+     */
+    postImage = async (userName, experimentId, image) => {
+      const folderPath = userName + '-' + experimentId + '/';
+      const imagePath = folderPath + image.name;
+      const rawImageRef = self.sieRawImagesBucketRef.child(imagePath);
+      rawImageRef.put(image).then(() => {
+        return true;
+      }).catch((error) => {
+        return false;
+      });
+    }
+
+    /**
+     * Get all existing self images in the messaging queue.
+     * @param {String} userName user name.
+     * @param {String} experimentId experiment id.
+     * @return {Image[]} array of self image files.
+     */
+    getImages = async (userName, experimentId) => {
+      // Creates a client; cache this for further use
+      pubSubClient = new PubSub();
+      // References an existing subscription
+      const subscription = pubSubClient.subscription(self.subscriptionName);
+
+      // Create an event handler to handle messages
+      const messageArray = [];
+      const messageHandler = (message) => {
+        messageArray.push(message.data);
+        // "Ack" (acknowledge receipt of) the message
+        message.ack();
+      };
+
+      // Listen for new messages until timeout is hit
+      subscription.on('message', messageHandler);
+
+      setTimeout(() => {
+        subscription.removeListener('message', messageHandler);
+      }, timeout * 1000);
+
+      // loop through the messages to locate those processed images
+      // message format: [userName]-[experimentId]-[status]
+      for (let index = 0; index < messageArray.length; index++) {
```

Comment:
will there be many messages sent? How does this work?

Reply:
So according to the design doc: "the cloud function will then put the newly processed images in another cloud storage bucket (sie-processed-images) and send a confirmation to firebase notifying users that the images have been processed." 

So there can be many messages sent in the message queue (of different users/experiments probably) when a pull request is triggered.

Thus we got to loop through the message queue to locate the specific user/experiment image processing status to get those processed images.

Reply:
hmm can we make it work so that we listen for messages and process one as soon as it's published? I must be misunderstanding things here. I am thinking of something like this:
- Cloud Run finishes image processing and publish a message
- This subscriber gets this message, parses it out and start the survey for the corresponding user

Oh, or are you getting several messages every few seconds or so then processing all of those in a loop?

Reply:
Yes, that is what I am doing right now, since we are now using the pull method to get messages asynchronously from the message queue, the subscriber cannot get the message once the publisher publishes the message but will pull when we trigger the getImages() method on client-side. Thereafter there can be multiple messages in the queue by the time we pull if multiple users are doing the experiment at the same time. So I am loop through the queue to find the relevant message for one specific user/experiment.